src/fetchers/tarball-fetcher.js

```diff
@@ -275,24 +290,43 @@ export default class TarballFetcher extends BaseFetcher {
               integrityValidateStream.pipe(fs.createWriteStream(tarballCachePath)).on('error', reject);
             }
 
-            integrityValidateStream.pipe(extractorStream).on('error', reject);
+            integrityValidateStream.pipe(extractorStream).on('error', async err => {
+              try {
+                // Cleanup before retrying
+                await this.removeDownloadedFiles();
+                if (this.validateError && this.validateError.code == 'EBADSIZE') {
+                  // Content-Length did not match the body size. Retry the request.
+                  if (!queueForRetry(err.message)) {
```

Comment:
This nested `if` blocks seems a bit untidy and confusing to me but I've also failed to come up with a clearer way. If anyone has suggestions, please share :)